irs/core/upload_handler.py

```diff
@@ -0,0 +1,235 @@
+        s3_access_key_id=CONFIG.s3_access_key_id,
+        s3_secret_access_key=CONFIG.s3_secret_access_key,
+    )
+    storage = S3ObjectStorage(config=config)
+
+    return storage.get_object_download_url(
+        bucket_id=CONFIG.bucket_id, object_id=object_id
+    )
+
+
+async def send_to_eks(
+    *, file_part: bytes, public_key: str, eks_url: str
+) -> Tuple[bytes, str, int]:
+    """Get encryption secret and file content offset from envelope"""
+    data = base64.b64encode(file_part).hex()
+    request_body = {"public_key": public_key, "file_part": data}
+    try:
+        response = requests.post(url=eks_url, json=request_body, timeout=60)
+    except requests.exceptions.RequestException as request_error:
+        raise exceptions.RequestFailedError(url=eks_url) from request_error
+
+    status_code = response.status_code
+    # implement httpyexpect error conversion
+    if status_code != 200:
+        spec = {
+            400: {
+                "malformedOrMissingEnvelopeError": exceptions.MalformedOrMissingEnvelope()
+            },
+            403: {"envelopeDecryptionError": exceptions.EnvelopeDecryptionError()},
+        }
+        ResponseExceptionTranslator(spec=spec).handle(response=response)
+        raise exceptions.BadResponseCodeError(url=eks_url, response_code=status_code)
+
+    body = response.json()
+    secret = base64.b64decode(codecs.decode(body["secret"], "hex"))
+    secret_id = body["secret_id"]
+    offset = body["offset"]
+
+    return secret, secret_id, offset
+
+
+async def compute_checksums(
+    *, download_url: str, secret: bytes, object_size: int, offset: int
+) -> Tuple[Tuple[List[str], List[str]], str]:
+    """Compute total unencrypted file checksum and encrypted part checksums"""
+    total_sha256_checksum = hashlib.sha256()
+    encrypted_md5_part_checksums = []
+    encrypted_sha256_part_checksums = []
+
+    # buffers to account for non part/cipher segment size aligned blocks
+    partial_chunk = b""
+
+    async for part in retrieve_parts(
+        url=download_url, object_size=object_size, offset=offset
+    ):
+        part = await part
+        md5sum, sha256sum = get_part_checksums(file_part=part)
+        encrypted_md5_part_checksums.append(md5sum)
+        encrypted_sha256_part_checksums.append(sha256sum)
+
+        if partial_chunk:
+            part = partial_chunk + part
+        chunks, incomplete_chunk = make_chunks(file_part=part)
+        partial_chunk = incomplete_chunk
+
+        for chunk in chunks:
+            decrypted = decrypt_block(ciphersegment=chunk, session_keys=[secret])
+            total_sha256_checksum.update(decrypted)
+
+    # process remaining data
+    if incomplete_chunk:
+        decrypted = decrypt_block(ciphersegment=incomplete_chunk, session_keys=[secret])
+        total_sha256_checksum.update(decrypted)
+
+    return (
+        encrypted_md5_part_checksums,
+        encrypted_sha256_part_checksums,
+    ), total_sha256_checksum.hexdigest()
```

Comment:
Maybe better return as flat tuple:
```suggestion
    return (
        encrypted_md5_part_checksums,
        encrypted_sha256_part_checksums, 
        total_sha256_checksum.hexdigest()
    )
```